### PR TITLE
Blackoilthermal fixes final

### PR DESCRIPTION
**Pull request by hnil**
-- added mixing model for thermal blackoil
-- TEMP is now set energy=enthalpy
-- THERMAL use enthalpy and energy different

This is draft pullrequest befor all commits are removed.

### CODE REVIEW

**Review by atgeirr — COMMENTED** [review]
I have now looked at the code, and think I understand the changes. I think it should be merged despite some uncertainty if this is exactly as it is or will need further changes, as it should be a clear improvement on what exists. Now testing on THERMAL and TEMP cases is needed to get this in, @vkip and @totto82 do you have a clear idea of the cases you think should be tested, and when you can do this testing?

**Comment by vkip** [comment]
>.. Now testing on THERMAL and TEMP cases is needed to get this in, @vkip and @totto82 do you have a clear idea of the cases you think should be tested, and when you can do this testing?

I have some TEMP cases that I can start looking at once the merge conflicts are resolved.

File: opm/input/eclipse/EclipseState/SimulationConfig/SimulationConfig.cpp [hunk 1]
```diff
@@ -107,6 +107,18 @@ namespace Opm {
...
                 || runspec.hasKeyword<ParserKeywords::TEMP>();
 
+            this->m_useEnthalpy = runspec.hasKeyword<ParserKeywords::THERMAL>();
+
+            if(runspec.hasKeyword<ParserKeywords::TEMP>()){
+                this->m_useEnthalpy = false;
+                if(runspec.hasKeyword<ParserKeywords::THERMAL>()){
+                    throw std::invalid_argument {
+                        "ERROR: In the RUNSPEC section the BOTH TEMP and THERMAL keyword "
+                        "only one should be spesified"
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/input/eclipse/EclipseState/SimulationConfig/SimulationConfig.cpp:117]: fixed

File: opm/material/fluidsystems/BlackOilFluidSystem.hpp [hunk 2]
```diff
@@ -1071,42 +1077,198 @@ class BlackOilFluidSystem : public BaseFluidSystem<Scalar, BlackOilFluidSystem<S
...
+    template <class FluidState, class LhsEval = typename FluidState::Scalar>
+    static LhsEval enthalpy(const FluidState& fluidState,
+                            unsigned phaseIdx,
+                            unsigned regionIdx)
+    {
+        // should preferably not be used values should be taken from intensive quantities fluid state.
+        const auto& p = decay<LhsEval>(fluidState.pressure(phaseIdx));
+        auto energy = internalEnergy<FluidState, LhsEval>(fluidState, phaseIdx, regionIdx);
+        if(!enthalpy_eq_energy_){
+            // used for simplified models
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/material/fluidsystems/BlackOilFluidSystem.hpp:1268]: fixed.

File: opm/material/fluidsystems/blackoilpvt/OilPvtThermal.hpp [hunk 3]
```diff
@@ -175,8 +176,9 @@ class OilPvtThermal
             return internalEnergyCurves_[regionIdx].eval(temperature, /*extrapolate=*/true);
         }
         else {
+            OpmLog::warning("Experimental code for jouleThomson: simulation will be slower");
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/material/fluidsystems/blackoilpvt/OilPvtThermal.hpp:179]: removed added comment

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this update_data failure_report please

**Comment by atgeirr** [comment]
jenkins build this opm-tests=4014 please

**Comment by akva2** [comment]
jenkins build this opm-tests=1226 please

**Comment by atgeirr** [comment]
Thanks for fixing my Jenkins messup, merging!

**Comment by bska** [comment]
> Thanks for fixing my Jenkins messup, merging!

Did you accidentally not merge the associate update to the reference solutions, OPM/opm-tests#1226?

**Comment by atgeirr** [comment]
> Did you accidentally not merge the associate update to the reference solutions, [OPM/opm-tests#1226](https://github.com/OPM/opm-tests/pull/1226)?

That is correct, my brain is not working today. Merged it now!

**Pull request merged**